cdap-proto/src/main/java/co/cask/cdap/proto/Id.java

```diff
@@ -200,4 +201,154 @@ public String toString() {
+    }
+
+    private NotificationFeed(String namespace, String category, String name, String description) {
+      Preconditions.checkArgument(namespace != null && !namespace.isEmpty(),
+                                  "Namespace value cannot be null or empty.");
+      Preconditions.checkArgument(category != null && !category.isEmpty(), "Category value cannot be null or empty.");
+      Preconditions.checkArgument(name != null && !name.isEmpty(), "Name value cannot be null or empty.");
+      Preconditions.checkArgument(isId(namespace) && isId(category) && isId(name),
+                                  "Namespace, category or name has a wrong format.");
+
+      this.namespace = Namespace.from(namespace);
+      this.category = category;
+      this.name = name;
+      this.id = String.format("%s.%s.%s", namespace, category, name);
+      this.description = description;
+    }
+
+    private boolean isId(String name) {
+      return CharMatcher.inRange('A', 'Z')
+        .or(CharMatcher.inRange('a', 'z'))
+        .or(CharMatcher.is('-'))
+        .or(CharMatcher.is('_'))
+        .or(CharMatcher.inRange('0', '9')).matchesAllOf(name);
+    }
+
+    public String getCategory() {
+      return category;
+    }
+
+    public String getId() {
+      return id;
+    }
+
+    public String getNamespaceId() {
+      return namespace.getId();
+    }
+
+    public String getName() {
+      return name;
+    }
+
+    public String getDescription() {
+      return description;
+    }
+
+    /**
+     * Builder used to build {@link NotificationFeed}.
+     */
+    public static final class Builder {
+      private String category;
+      private String name;
+      private String namespaceId;
+      private String description;
+
+      public Builder() {
+        // No-op
+      }
+
+      public Builder(NotificationFeed feed) {
+        this.namespaceId = feed.getNamespaceId();
+        this.category = feed.getCategory();
+        this.name = feed.getName();
+        this.description = feed.getDescription();
+      }
+
+      public Builder setName(final String name) {
+        this.name = name;
+        return this;
+      }
+
+      public Builder setNamespaceId(final String namespace) {
+        this.namespaceId = namespace;
+        return this;
+      }
+
+      public Builder setDescription(final String description) {
+        this.description = description;
+        return this;
+      }
+
+      public Builder setCategory(final String category) {
+        this.category = category;
+        return this;
+      }
+
+      /**
+       * @return a {@link NotificationFeed} object containing all the fields set in the builder.
+       * @throws IllegalArgumentException if the namespaceId, category or name is invalid.
+       */
+      public NotificationFeed build() {
+        return new NotificationFeed(namespaceId, category, name, description);
+      }
+    }
+
+    @Override
+    public String toString() {
```

Comment:
I can see that the ``id`` here contains namespace, category and name, but better to use all fields in ``toStringHelper()`` so the formatting is consistent?